tensorflow/core/ir/ops.cc

```diff
@@ -194,8 +194,15 @@ void TFGraphDialect::printCustomTfOp(Operation *op,
   }
 
   // Handles the optional "device" and "name" attribute.
-  ArrayRef<llvm::StringRef> keywords{"_mlir_device", "_mlir_assigned_device",
```

Comment:
Do you see an issue when running under the ASAN?

I'm wondering since literals, e.g,  "_mlir_device" have static storage duration and live the whole duration of the program.
So, taking StringRef from a string literal should be completely safe.

ArrayRef<...> keywords{"..", ".."} get's constructed from the initializer_list<llvm::StringRef> (https://llvm.org/doxygen/ArrayRef_8h_source.html#l00114)
which is safe as well, make copies internally.

Reply:
What I see is that before https://github.com/tensorflow/tensorflow/blob/4a265a6f3a8ea441e6135da03aafa773bbce5505/tensorflow/core/ir/ops.cc#L202 the array that ArrayRef keywords references is OK, but after that line, the first entry in that array has been overwritten. Looking at the assembler I can see the write to the memory location of that array which appears to be being used as temporary storage. So my conclusion was that the storage for the array itself was only temporarily allocated and hence available to the compiler for re-use. Depending on the optimization level in use, this could either lead to a crash or else just incorrect behaviour.
Now it may be overkill to create Strings for the string literals for the StringRefs to refer to, but something certainly needs to be done to extend the life of the array of those StringRefs. The compiler believes that memory is fair game.

Reply:
Interesting, I see. Could you check if this is sufficient?
```
  std::array<llvm::StringRef, 3> keywords = {"_mlir_device", "_mlir_assigned_device", "_mlir_name"};
  ```

Reply:
Yes that is sufficient.

Reply:
Sounds good!